File: src/kernel_builder/builder.py
"""
FORGED — Android Kernel Builder
Copyright (c) 2026 vxyzview. Made with love.

Kernel build engine — orchestrates make invocations.
"""

from __future__ import annotations

import multiprocessing
import os
import shutil
import subprocess
import time
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path

from .config import BuildConfig
from .toolchain_manager import auto_setup_toolchain, clone_kernel_source, toolchain_clang_path

# ── LLVM binutils tool names ──────────────────────────────────────────────────
#
# These are passed as explicit make variables when use_llvm_binutils=True,
# replacing the GNU equivalents and removing any dependency on GCC.

LLVM_BINUTILS: dict[str, str] = {
    "LD": "ld.lld",
    "AR": "llvm-ar",
    "NM": "llvm-nm",
    "OBJCOPY": "llvm-objcopy",
    "OBJDUMP": "llvm-objdump",
    "READELF": "llvm-readelf",
    "STRIP": "llvm-strip",
}


# ── Result types ──────────────────────────────────────────────────────────────


@dataclass
class BuildStep:
    name: str
    command: list[str]
    env: dict


@dataclass
class BuildResult:
    success: bool
    step: str
    duration: float
    output: str
    error: str = ""


# ── Helpers ───────────────────────────────────────────────────────────────────


def _resolve_jobs(requested: int) -> int:
    if requested > 0:
        return requested
    return max(1, multiprocessing.cpu_count())


def find_ccache() -> Path | None:
    """Return the absolute path to the ccache binary, or None if not found.

    The binary is looked up via PATH so it works whether ccache is installed
    system-wide (``/usr/bin/ccache``) or in a custom prefix.
    """
    found = shutil.which("ccache")
    return Path(found) if found else None


def _build_env(cfg: BuildConfig) -> dict:
    """Build the subprocess environment for all make invocations.

    Always uses Clang as CC.  When use_llvm_binutils is True, the LLVM
    equivalents of all binutils tools are exported so no GCC installation
    is required on the host.

    Cross-compiler variables exported
    ───────────────────────────────────
    CROSS_COMPILE       — arm64 GNU prefix (always exported).
    CROSS_COMPILE_ARM32 — explicit arm32 prefix for modern Android kernel trees
                          (exported when toolchain.cross_compile_arm32 is non-empty).
                          Leave the field empty in your config to suppress it.

    When ccache is enabled the following CCACHE_* variables are exported:

      CCACHE_DIR        — storage directory (when cfg.ccache.dir is set)
      CCACHE_MAXSIZE    — maximum cache size (e.g. "5G")
      CCACHE_SLOPPINESS — comma-separated flags tuned for kernel builds
      CCACHE_COMPRESS   — "true" / "false"
      CCACHE_BASEDIR    — base dir for relocatable entries (defaults to source)

    The CC make variable is *not* modified here; the ``ccache`` prefix is
    injected in ``_make_base()`` so it appears on the make command line too,
    which is required for ``ccache`` to intercept sub-make invocations
    (e.g. modpost).

    extra_path entries are expanded with os.path.expanduser /
    os.path.expandvars so that values like ~/toolchains/clang/bin work
    correctly instead of being passed through literally as '~'.

    extra_env is applied last, after all built-in variables, so it can
    override any variable when needed (e.g. to set KCPPFLAGS or KBUILD_VERBOSE).
    """
    env = os.environ.copy()

    tc = cfg.toolchain
    expanded_paths = [
        os.path.expandvars(os.path.expanduser(str(p)))
        for p in tc.extra_path
        if p
    ]
    extra_paths = ":".join(expanded_paths)
    if extra_paths:
        env["PATH"] = extra_paths + ":" + env.get("PATH", "")

    env.update(
        {
            "ARCH": cfg.arch,
            "SUBARCH": cfg.subarch,
            "CC": tc.cc,
            "CROSS_COMPILE": tc.cross_compile,
            "CLANG_TRIPLE": tc.clang_triple,
            "KBUILD_BUILD_USER": cfg.kbuild_build_user,
            "KBUILD_BUILD_HOST": cfg.kbuild_build_host,
        }
    )

    # CROSS_COMPILE_ARM32 — explicit arm32 prefix (modern Android kernel trees).
    # Omitted entirely when the field is empty so kernels that don't recognise
    # the variable are not affected.
    if tc.cross_compile_arm32:
        env["CROSS_COMPILE_ARM32"] = tc.cross_compile_arm32

    if tc.use_llvm_binutils:
        env.update(LLVM_BINUTILS)

    if cfg.localversion:
        env["LOCALVERSION"] = cfg.localversion

    # ── ccache environment ────────────────────────────────────────────────────
    if cfg.ccache.enabled:
        cc = cfg.ccache

        if cc.dir:
            env["CCACHE_DIR"] = os.path.expandvars(os.path.expanduser(cc.dir))

        env["CCACHE_MAXSIZE"] = cc.max_size
        env["CCACHE_SLOPPINESS"] = cc.sloppiness
        env["CCACHE_COMPRESS"] = "true" if cc.compress else "false"

        # CCACHE_BASEDIR helps ccache share cache entries across builds that
        # use different absolute paths (e.g. CI checkouts in /tmp/buildNNN).
        # Default to the kernel source dir when the user did not set basedir.
        raw_basedir = cc.basedir or cfg.kernel_source
        if raw_basedir:
            env["CCACHE_BASEDIR"] = os.path.expandvars(
                os.path.expanduser(str(raw_basedir))
            )

    # ── User-supplied arbitrary environment variables (applied last) ──────────
    # extra_env overrides any of the built-in variables above when they share
    # the same key.  Tilde and $VAR references are expanded for convenience.
    for key, value in cfg.extra_env.items():
        env[key] = os.path.expandvars(os.path.expanduser(value))

    return env


def _make_base(cfg: BuildConfig, jobs: int) -> list[str]:
    """Construct the common make command prefix used by every build step.

    Always emits CC/CLANG_TRIPLE/CROSS_COMPILE/SUBARCH flags.  When
    use_llvm_binutils is True, all LLVM binutils overrides are appended so
    the kernel Makefile never falls back to the GNU tools.

    Cross-compiler make variables
    ──────────────────────────────
    CROSS_COMPILE       — always included (arm64 prefix).
    CROSS_COMPILE_ARM32 — included when non-empty (modern arm32 prefix for
                          Android / GKI trees).  Omitted when the field is
                          empty so the kernel Makefile is not affected.

    When ccache is enabled, the CC make variable is set to
    ``"ccache <compiler>"`` (e.g. ``CC=ccache clang``).  This is separate
    from the CCACHE_* environment variables set in ``_build_env()`` — both
    are required so that ccache intercepts the top-level compiler call *and*
    any recursive make invocations that pass CC explicitly.

    SUBARCH is included on the make command line in addition to the
    environment so that kernel Makefiles that expect it as a variable
    (not just an env export) behave correctly.

    extra_make_flags are appended verbatim at the end, after all built-in
    variables, so they can extend or override anything above (e.g. LLVM=1).
    """
    tc = cfg.toolchain

    # Prefix compiler with ccache when enabled.
    cc_value = f"ccache {tc.cc}" if cfg.ccache.enabled else tc.cc

    cmd: list[str] = [
        "make",
        f"-j{jobs}",
        f"O={cfg.output_dir}",
        f"ARCH={cfg.arch}",
        f"SUBARCH={cfg.subarch}",
        f"CC={cc_value}",
        f"CLANG_TRIPLE={tc.clang_triple}",
        f"CROSS_COMPILE={tc.cross_compile}",
    ]

    # CROSS_COMPILE_ARM32 — modern arm32 prefix (Android / GKI kernel trees).
    # Omitted entirely when empty to avoid confusing kernels that don't expect it.
    if tc.cross_compile_arm32:
        cmd.append(f"CROSS_COMPILE_ARM32={tc.cross_compile_arm32}")

    if tc.use_llvm_binutils:
        cmd.extend(f"{var}={tool}" for var, tool in LLVM_BINUTILS.items())

    if cfg.lto == "thin":
        cmd.append("LTO=thin")
    elif cfg.lto == "full":
        cmd.append("LTO=full")

    # Append user-supplied flags last so they can extend or override anything.
    cmd.extend(cfg.extra_make_flags)
    return cmd


# ── Build steps ───────────────────────────────────────────────────────────────


def _step_mrproper(cfg: BuildConfig, jobs: int, env: dict) -> BuildStep:
    return BuildStep(
        name="mrproper",
        command=_make_base(cfg, jobs) + ["mrproper"],
        env=env,
    )


def _step_defconfig(cfg: BuildConfig, jobs: int, env: dict) -> BuildStep:
    return BuildStep(
        name="defconfig",
        command=_make_base(cfg, jobs) + [cfg.kernel_defconfig],
        env=env,
    )


def _step_build(cfg: BuildConfig, jobs: int, env: dict) -> BuildStep:
    return BuildStep(
        name="build",
        command=_make_base(cfg, jobs),
        env=env,
    )


# ── Runner ────────────────────────────────────────────────────────────────────


class KernelBuilder:
    """Orchestrate kernel build steps, yielding BuildResult for each."""

    def __init__(
        self,
        cfg: BuildConfig,
        progress_callback: Callable[[str], None] | None = None,
    ) -> None:
        self.cfg = cfg
        self.jobs = _resolve_jobs(cfg.jobs)

        # ── Auto-clone kernel source from git URL ─────────────────────────────
        # NOTE: cfg.kernel_source is mutated in-place when kernel_source_url is
        # set, so the caller's BuildConfig object will reflect the cloned path.
        if cfg.kernel_source_url:
            dest = Path(cfg.kernel_source) if cfg.kernel_source else Path.cwd() / "kernel"
            clone_kernel_source(
                cfg.kernel_source_url,
                dest,
                branch=cfg.kernel_source_branch,
                depth=cfg.kernel_source_depth,
                progress=progress_callback,
            )
            cfg.kernel_source = str(dest)

        self.source_dir = Path(cfg.kernel_source).resolve()
        self.output_dir = self.source_dir / cfg.output_dir

        # ── Auto-clone AOSP Clang and set arm64/arm PATH ──────────────────────
        if cfg.auto_setup_toolchain and cfg.toolchain.auto_clone:
            toolchain_base = (
                Path(cfg.toolchain_dir) if cfg.toolchain_dir else None
            )
            auto_setup_toolchain(
                cfg,
                toolchain_base=toolchain_base,
                install_cross_compilers=False,
                progress=progress_callback,
            )

    def validate_toolchain(self) -> list[str]:
        """Return a list of warning strings about missing toolchain components.

        Checks for the configured Clang binary and, when ccache is enabled,
        verifies that the ``ccache`` binary is available on PATH.  An empty
        list means everything looks fine.
        """
        issues: list[str] = []

        clang = toolchain_clang_path(self.cfg)
        if clang is None:
            issues.append(
                f"Clang binary '{self.cfg.toolchain.cc}' not found in "
                "extra_path or $PATH.  Run 'kernel-builder setup-toolchain' "
                "or set toolchain.extra_path in your build config."
            )

        # ── ccache binary check ───────────────────────────────────────────────
        if self.cfg.ccache.enabled and find_ccache() is None:
            issues.append(
                "ccache is enabled but the 'ccache' binary was not found in "
                "$PATH.  Install it with:  sudo apt install ccache\n"
                "Alternatively, disable ccache in your build config."
            )

        return issues

    # ── Public API ────────────────────────────────────────────────────────────

    def steps(self, clean: bool = True) -> list[BuildStep]:
        env = _build_env(self.cfg)
        step_list: list[BuildStep] = []
        if clean:
            step_list.append(_step_mrproper(self.cfg, self.jobs, env))
        step_list.append(_step_defconfig(self.cfg, self.jobs, env))
        step_list.append(_step_build(self.cfg, self.jobs, env))
        return step_list

    def run_step(
        self,
        step: BuildStep,
        line_callback: Callable[[str], None] | None = None,
    ) -> BuildResult:
        """Execute one build step, streaming stdout/stderr line-by-line."""
        start = time.monotonic()
        combined_lines: list[str] = []

        try:
            process = subprocess.Popen(
                step.command,
                cwd=str(self.source_dir),
                env=step.env,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                bufsize=1,
            )
            if process.stdout is None:  # pragma: no cover — Popen with PIPE always sets stdout
                raise RuntimeError("subprocess stdout is None; this should never happen")
            for line in process.stdout:
                stripped = line.rstrip()
                combined_lines.append(stripped)
                if line_callback:
                    line_callback(stripped)
            process.wait()
            duration = time.monotonic() - start
            success = process.returncode == 0
            return BuildResult(
                success=success,
                step=step.name,
                duration=duration,
                output="\n".join(combined_lines),
                error="" if success else f"Exit code {process.returncode}",
            )
        except FileNotFoundError as exc:
            duration = time.monotonic() - start
            return BuildResult(
                success=False,
                step=step.name,
                duration=duration,
                output="",
                error=f"Command not found: {exc}",
            )
        except Exception as exc:  # noqa: BLE001
            duration = time.monotonic() - start
            return BuildResult(
                success=False,
                step=step.name,
                duration=duration,
                output="\n".join(combined_lines),
                error=str(exc),
            )

    def find_kernel_image(self) -> Path | None:
        """Locate the compiled kernel image (Image, Image.gz, zImage, etc.)."""
        search_names = [
            "Image.gz-dtb",
            "Image-dtb",
            "Image.gz",
            "Image",
            "zImage-dtb",
            "zImage",
            "dtb.img",
        ]
        for name in search_names:
            candidate = self.output_dir / "arch" / self.cfg.arch / "boot" / name
            if candidate.exists():
                return candidate
        return None

    def find_dtb_files(self) -> list[Path]:
        """Collect unique .dtb files from output.

        Results are deduplicated via a seen-set so that files matched
        by both the recursive and top-level glob patterns are not returned
        twice.
        """
        boot_dir = self.output_dir / "arch" / self.cfg.arch / "boot"
        seen: set[Path] = set()
        dtbs: list[Path] = []
        for pattern in ("dts/**/*.dtb", "*.dtb"):
            for dtb in boot_dir.glob(pattern):
                resolved = dtb.resolve()
                if resolved not in seen:
                    seen.add(resolved)
                    dtbs.append(dtb)
        return dtbs

    def find_modules(self) -> list[Path]:
        """Collect compiled kernel modules (.ko files)."""
        return list(self.output_dir.rglob("*.ko"))

File: internal/builder/builder.go
// Package builder implements the FORGED kernel build engine: make invocation
// preparation (environment + command prefix) and step orchestration.
//
// FORGED — Android Kernel Builder
// Copyright (c) 2026 vxyzview. Made with love.
package builder

import (
	"bufio"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/vxyzview/forged/internal/config"
	"github.com/vxyzview/forged/internal/toolchain"
)

// LLVM binutils: passed as explicit make variables (and env) when
// use_llvm_binutils is true, removing any dependency on GCC.
var LLVMBinutils = map[string]string{
	"LD":      "ld.lld",
	"AR":      "llvm-ar",
	"NM":      "llvm-nm",
	"OBJCOPY": "llvm-objcopy",
	"OBJDUMP": "llvm-objdump",
	"READELF": "llvm-readelf",
	"STRIP":   "llvm-strip",
}

// Ordered keys so make/env output is deterministic.
var llvmBinutilsOrder = []string{"LD", "AR", "NM", "OBJCOPY", "OBJDUMP", "READELF", "STRIP"}

// BuildStep is one make invocation (mrproper / defconfig / build).
type BuildStep struct {
	Name    string
	Command []string
	Env     map[string]string
}

// BuildResult is the outcome of executing a BuildStep.
type BuildResult struct {
	Success  bool
	Step     string
	Duration float64
	Output   string
	Error    string
}

// LineCallback receives streamed output lines.
type LineCallback func(line string)

// ProgressCallback receives progress/log lines.
type ProgressCallback func(line string)

// ResolveJobs returns requested, or the CPU count when requested <= 0.
func ResolveJobs(requested int) int {
	if requested > 0 {
		return requested
	}
	if n := runtime.NumCPU(); n > 0 {
		return n
	}
	return 1
}

// FindCcache returns the absolute ccache path, or "".
func FindCcache() string {
	return toolchain.Which("ccache")
}

// BuildEnv builds the subprocess environment for every make invocation.
//
// Always uses Clang as CC. When use_llvm_binutils is true, LLVM binutils are
// exported. CROSS_COMPILE_ARM32 is exported when non-empty. When ccache is
// enabled, CCACHE_* vars are exported (the "ccache" prefix on CC is injected
// separately in MakeBase). ExtraEnv is applied last so it can override any
// built-in variable.
func BuildEnv(cfg *config.BuildConfig) map[string]string {
	env := map[string]string{}
	for _, kv := range os.Environ() {
		if i := strings.IndexByte(kv, '='); i >= 0 {
			env[kv[:i]] = kv[i+1:]
		}
	}

	tc := &cfg.Toolchain

	var expanded []string
	for _, p := range tc.ExtraPath {
		if p == "" {
			continue
		}
		expanded = append(expanded, config.ExpandPath(p))
	}
	if joined := strings.Join(expanded, ":"); joined != "" {
		if base := env["PATH"]; base != "" {
			env["PATH"] = joined + ":" + base
		} else {
			env["PATH"] = joined
		}
	}

	env["ARCH"] = cfg.Arch
	env["SUBARCH"] = cfg.Subarch
	env["CC"] = tc.CC
	env["CROSS_COMPILE"] = tc.CrossCompile
	env["CLANG_TRIPLE"] = tc.ClangTriple
	env["KBUILD_BUILD_USER"] = cfg.KBUILDBuildUser
	env["KBUILD_BUILD_HOST"] = cfg.KBUILDBuildHost

	if tc.CrossCompileArm32 != "" {
		env["CROSS_COMPILE_ARM32"] = tc.CrossCompileArm32
	}

	if tc.UseLLVMBinutils {
		for _, k := range llvmBinutilsOrder {
			env[k] = LLVMBinutils[k]
		}
	}

	if cfg.Localversion != "" {
		env["LOCALVERSION"] = cfg.Localversion
	}

	if cfg.Ccache.Enabled {
		cc := &cfg.Ccache
		if cc.Dir != "" {
			env["CCACHE_DIR"] = config.ExpandPath(cc.Dir)
		}
		env["CCACHE_MAXSIZE"] = cc.MaxSize
		env["CCACHE_SLOPPINESS"] = cc.Sloppiness
		if cc.Compress {
			env["CCACHE_COMPRESS"] = "true"
		} else {
			env["CCACHE_COMPRESS"] = "false"
		}
		rawBasedir := cc.Basedir
		if rawBasedir == "" {
			rawBasedir = cfg.KernelSource
		}
		if rawBasedir != "" {
			env["CCACHE_BASEDIR"] = config.ExpandPath(rawBasedir)
		}
	}

	for k, v := range cfg.ExtraEnv {
		env[k] = config.ExpandPath(v)
	}

	return env
}

// MakeBase constructs the common make command prefix for every build step.
//
// Always emits CC/CLANG_TRIPLE/CROSS_COMPILE/SUBARCH. When ccache is enabled,
// CC is set to "ccache <compiler>". ExtraMakeFlags are appended verbatim last.
func MakeBase(cfg *config.BuildConfig, jobs int) []string {
	tc := &cfg.Toolchain

	ccValue := tc.CC
	if cfg.Ccache.Enabled {
		ccValue = "ccache " + tc.CC
	}

	cmd := []string{
		"make",
		fmt.Sprintf("-j%d", jobs),
		"O=" + cfg.OutputDir,
		"ARCH=" + cfg.Arch,
		"SUBARCH=" + cfg.Subarch,
		"CC=" + ccValue,
		"CLANG_TRIPLE=" + tc.ClangTriple,
		"CROSS_COMPILE=" + tc.CrossCompile,
	}

	if tc.CrossCompileArm32 != "" {
		cmd = append(cmd, "CROSS_COMPILE_ARM32="+tc.CrossCompileArm32)
	}

	if tc.UseLLVMBinutils {
		for _, k := range llvmBinutilsOrder {
			cmd = append(cmd, k+"="+LLVMBinutils[k])
		}
	}

	switch cfg.LTO {
	case "thin":
		cmd = append(cmd, "LTO=thin")
	case "full":
		cmd = append(cmd, "LTO=full")
	}

	cmd = append(cmd, cfg.ExtraMakeFlags...)
	return cmd
}

// KernelBuilder orchestrates kernel build steps.
type KernelBuilder struct {
	Cfg       *config.BuildConfig
	Jobs      int
	SourceDir string
	OutputDir string
}

// New creates a KernelBuilder. When cfg.KernelSourceURL is set the kernel
// source is cloned first (mutating cfg.KernelSource in place), then the
// toolchain is auto-set-up when enabled. Pass nil progress to silence logs.
func New(ctx context.Context, cfg *config.BuildConfig, progress ProgressCallback) (*KernelBuilder, error) {
	jobs := ResolveJobs(cfg.Jobs)

	if cfg.KernelSourceURL != "" {
		dest := cfg.KernelSource
		if dest == "" {
			cwd, err := os.Getwd()
			if err != nil {
				return nil, err
			}
			dest = filepath.Join(cwd, "kernel")
		}
		var cb toolchain.Progress
		if progress != nil {
			cb = toolchain.Progress(progress)
		}
		cloned, err := toolchain.CloneKernelSource(ctx, cfg.KernelSourceURL, dest, cfg.KernelSourceBranch, cfg.KernelSourceDepth, cb)
		if err != nil {
			return nil, err
		}
		cfg.KernelSource = cloned
	}

	sourceDir, err := filepath.Abs(cfg.KernelSource)
	if err != nil {
		return nil, err
	}

	if cfg.AutoSetupToolchain && cfg.Toolchain.AutoClone {
		var cb toolchain.Progress
		if progress != nil {
			cb = toolchain.Progress(progress)
		}
		if _, err := toolchain.AutoSetupToolchain(ctx, cfg, cfg.ToolchainDir, false, cb); err != nil {
			return nil, err
		}
	}

	return &KernelBuilder{
		Cfg:       cfg,
		Jobs:      jobs,
		SourceDir: sourceDir,
		OutputDir: filepath.Join(sourceDir, cfg.OutputDir),
	}, nil
}

// ValidateToolchain returns warnings about missing toolchain components.
func (b *KernelBuilder) ValidateToolchain() []string {
	var issues []string

	if toolchain.ToolchainClangPath(b.Cfg) == "" {
		issues = append(issues,
			fmt.Sprintf("Clang binary '%s' not found in extra_path or $PATH.  Run 'forged setup-toolchain' or set toolchain.extra_path in your build config.", b.Cfg.Toolchain.CC))
	}
	if b.Cfg.Ccache.Enabled && FindCcache() == "" {
		issues = append(issues,
			"ccache is enabled but the 'ccache' binary was not found in $PATH.  Install it with:  sudo apt install ccache\nAlternatively, disable ccache in your build config.")
	}
	return issues
}

// Steps returns the ordered build steps (mrproper optionally, then defconfig
// and build).
func (b *KernelBuilder) Steps(clean bool) []BuildStep {
	env := BuildEnv(b.Cfg)
	var steps []BuildStep
	if clean {
		steps = append(steps, BuildStep{Name: "mrproper", Command: append(MakeBase(b.Cfg, b.Jobs), "mrproper"), Env: env})
	}
	steps = append(steps,
		BuildStep{Name: "defconfig", Command: append(MakeBase(b.Cfg, b.Jobs), b.Cfg.KernelDefconfig), Env: env},
		BuildStep{Name: "build", Command: MakeBase(b.Cfg, b.Jobs), Env: env},
	)
	return steps
}

// RunStep executes one build step, streaming stdout+stderr line by line.
func (b *KernelBuilder) RunStep(ctx context.Context, step BuildStep, cb LineCallback) BuildResult {
	start := time.Now()
	var combined []string

	envList := envMapToList(step.Env)

	c := exec.CommandContext(ctx, step.Command[0], step.Command[1:]...)
	c.Dir = b.SourceDir
	c.Env = envList

	stdout, err := c.StdoutPipe()
	if err != nil {
		return BuildResult{Success: false, Step: step.Name, Duration: time.Since(start).Seconds(), Error: err.Error()}
	}
	c.Stderr = c.Stdout

	if err := c.Start(); err != nil {
		msg := err.Error()
		// exec.Error for missing binary reports "executable file not found".
		return BuildResult{Success: false, Step: step.Name, Duration: time.Since(start).Seconds(), Error: msg}
	}

	scanner := bufio.NewScanner(stdout)
	scanner.Buffer(make([]byte, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimRight(scanner.Text(), "\r")
		combined = append(combined, line)
		if cb != nil {
			cb(line)
		}
	}
	waitErr := c.Wait()
	duration := time.Since(start).Seconds()
	output := strings.Join(combined, "\n")

	if waitErr != nil {
		if exitErr, ok := waitErr.(*exec.ExitError); ok {
			return BuildResult{Success: false, Step: step.Name, Duration: duration, Output: output, Error: fmt.Sprintf("Exit code %d", exitErr.ExitCode())}
		}
		return BuildResult{Success: false, Step: step.Name, Duration: duration, Output: output, Error: waitErr.Error()}
	}
	return BuildResult{Success: true, Step: step.Name, Duration: duration, Output: output}
}

// searchNames is the kernel image search priority.
var searchNames = []string{
	"Image.gz-dtb",
	"Image-dtb",
	"Image.gz",
	"Image",
	"zImage-dtb",
	"zImage",
	"dtb.img",
}

// FindKernelImage locates the compiled kernel image, or "" when absent.
func (b *KernelBuilder) FindKernelImage() string {
	for _, name := range searchNames {
		candidate := filepath.Join(b.OutputDir, "arch", b.Cfg.Arch, "boot", name)
		if fi, err := os.Stat(candidate); err == nil && !fi.IsDir() {
			return candidate
		}
	}
	return ""
}

// FindDTBFiles collects unique .dtb files from the output tree.
func (b *KernelBuilder) FindDTBFiles() []string {
	bootDir := filepath.Join(b.OutputDir, "arch", b.Cfg.Arch, "boot")
	seen := map[string]bool{}
	var dtbs []string

	// Recursive dts/**/*.dtb first, then top-level *.dtb — deduplicated.
	var walk func(dir string, top bool)
	walk = func(dir string, top bool) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		for _, e := range entries {
			p := filepath.Join(dir, e.Name())
			if e.IsDir() {
				if dir == bootDir {
					walk(p, false)
				}
				continue
			}
			if !strings.HasSuffix(e.Name(), ".dtb") {
				continue
			}
			if top && dir != bootDir {
				continue
			}
			// Only include top-level hits on the first pass; recursive pass
			// covers nested dirs.
			if !top && dir == bootDir {
				continue
			}
			resolved, err := filepath.EvalSymlinks(p)
			if err != nil {
				resolved = p
			}
			if !seen[resolved] {
				seen[resolved] = true
				dtbs = append(dtbs, p)
			}
		}
	}
	// Recursive walk skipping top-level files:
	var recWalk func(dir string, isTop bool)
	recWalk = func(dir string, isTop bool) {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return
		}
		for _, e := range entries {
			p := filepath.Join(dir, e.Name())
			if e.IsDir() {
				recWalk(p, false)
				continue
			}
			if isTop {
				continue
			}
			if !strings.HasSuffix(e.Name(), ".dtb") {
				continue
			}
			resolved, err := filepath.EvalSymlinks(p)
			if err != nil {
				resolved = p
			}
			if !seen[resolved] {
				seen[resolved] = true
				dtbs = append(dtbs, p)
			}
		}
	}
	recWalk(bootDir, true)
	// Top-level *.dtb:
	if entries, err := os.ReadDir(bootDir); err == nil {
		for _, e := range entries {
			if e.IsDir() || !strings.HasSuffix(e.Name(), ".dtb") {
				continue
			}
			p := filepath.Join(bootDir, e.Name())
			resolved, err := filepath.EvalSymlinks(p)
			if err != nil {
				resolved = p
			}
			if !seen[resolved] {
				seen[resolved] = true
				dtbs = append(dtbs, p)
			}
		}
	}
	_ = walk
	return dtbs
}

// FindModules collects compiled kernel modules (.ko files).
func (b *KernelBuilder) FindModules() []string {
	var mods []string
	_ = filepath.Walk(b.OutputDir, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if !info.IsDir() && strings.HasSuffix(info.Name(), ".ko") {
			mods = append(mods, p)
		}
		return nil
	})
	if mods == nil {
		return []string{}
	}
	return mods
}

func envMapToList(env map[string]string) []string {
	out := make([]string, 0, len(env))
	for k, v := range env {
		out = append(out, k+"="+v)
	}
	return out
}
